elide-graphql/src/main/java/com/yahoo/elide/graphql/GraphQLEntityProjectionMaker.java

```diff
@@ -0,0 +1,410 @@
+                FragmentDefinition fragmentDefinition = (FragmentDefinition) definition;
+                fragmentDefinitions.put(fragmentDefinition.getName(), fragmentDefinition);
+            } else {
+                throw new InvalidEntityBodyException(
+                        String.format("Unsupported definition type {%s}.", definition.getClass()));
+            }
+        });
+
+        // make sure there is no fragment loop and undefined fragments in fragment definitions
+        final Set<String> fragmentNames = new HashSet<>();
+        fragmentDefinitions.values()
+                .forEach(fragmentDefinition -> validateFragment(fragmentDefinition, fragmentNames));
+
+        // make sure all fragments in projections are defined
+        if (fragmentToProjections.keySet().stream()
+                .anyMatch(fragmentName -> !fragmentDefinitions.containsKey(fragmentName))) {
+            throw new InvalidEntityBodyException("Unknown fragments {" +
+                    fragmentToProjections.keySet().stream()
+                            .filter(fragmentName -> !fragmentDefinitions.containsKey(fragmentName))
+                            .collect(Collectors.joining(",")) +
+                    "}.");
+        }
+
+        // as there is no loop in the fragment definition, this loop would eventually end
+        while(!fragmentToProjections.isEmpty()) {
+            fragmentDefinitions.forEach((fragmentName, fragmentDefinition) -> {
+                if (fragmentToProjections.containsKey(fragmentName)) {
+                    List<Selection> fragmentFields = fragmentDefinition.getSelectionSet().getSelections();
+                    List<EntityProjection> projections = fragmentToProjections.remove(fragmentName);
+
+                    // the entity type of the projection must be the same type in the fragment type condition
+                    String fragmentTypeName = fragmentDefinition.getTypeCondition().getName();
+
+                    // add fields into projections, if new fragments are referenced, they would be added into the map
+                    projections.stream()
+                            .filter(projection -> projection.getName().equals(fragmentTypeName))
+                            .forEach(projection -> fragmentFields.forEach(field -> addField(field, projection)));
+                }
+            });
+        }
+
+        return rootProjections;
+    }
+
+    /**
+     * Root projection would be an operation applied on an entity class. There should be only one selection, which
+     * is the entity, in the selection set. The EntityProjection tree would be constructed recursively to add all
+     * child projections.
+     */
+    private void addRootProjections(SelectionSet selectionSet) {
+        List<Selection> selections = selectionSet.getSelections();
+
+        if (selections.size() != 1) {
+            throw new InvalidEntityBodyException("Can't select multiple entities in graphQL QUERY operation.");
```

Comment:
I don't think this is true.  YOu can create multiple roots - and there can be aliases for the same root asked for twice in the same query.  There is an IT tests for this.